File: py/GameSession.py

import engine

class GameSession:

    def __init__(self, wordsFilePath: str, foundWordCallback, gameOverCallback, startingHearts: int = 3, maxHearts: int = 3):
        
        self.letterWeights = engine.generateDefaultWeights()
        self.acceptedWords = engine.getWordsFromFile(wordsFilePath)
        self.maxHearts = maxHearts

        self.foundWordCallback = foundWordCallback
        self.gameOverCallback = gameOverCallback

        self.turnsPlayed = 0
        self.lastFoundWord = None
        self.lastQueriedSubstring = None
        self.hearts = startingHearts
        self.heartRefills = 0
        self.lostHearts = 0


    def allLettersAreUsed(self) -> bool:

        return not any(self.letterWeights.values())

    def resetLetterWeights(self):

        self.letterWeights = engine.generateDefaultWeights()
    
    def averageTurnsPerRefill(self) -> float:

        return self.turnsPlayed / self.heartRefills
    
    def queryOnSubstring(self, substring: str) -> str:
        
        self.lastQueriedSubstring = substring
        self.turnsPlayed += 1
        self.lastFoundWord = engine.playTurn(substring, self.acceptedWords, self.letterWeights, self.foundWordCallback)

        if self.lastFoundWord == None:

            self.hearts -= 1
            self.lostHearts =+1

            if self.hearts == 0:
                self.gameOverCallback()

            return None

        if self.allLettersAreUsed():

            self.resetLetterWeights()
            self.heartRefills += 1
            self.hearts = min(self.hearts + 1, self.maxHearts)

        
        return self.lastFoundWord

    def toDict(self) -> dict:

        return {
            "turn": self.turnsPlayed,
            "lastQueriedSubstring": self.lastQueriedSubstring,
            "lastFoundWord": self.lastFoundWord,
            "hearts": self.hearts,
            "refillCount": self.heartRefills,
            "unusedLetters": [letter for letter in self.letterWeights if self.letterWeights[letter] > 0],
       }

File: cpp/GameSession.cpp

#include "GameSession.hpp"

GameSession::GameSession(
    std::filesystem::path wordsFilePath,
    std::function<void(const std::string&)> foundWordCallback,
    std::function<void()> gameOverCallback,
    int startingHearts,
    int maxHearts
) : foundWordCallback(foundWordCallback), gameOverCallback(gameOverCallback), hearts(startingHearts), maxHearts(maxHearts), turnsPlayed(0), heartRefills(0), lostHearts(0)
{
    letterWeights = Engine::generateDefaultWeights();
    acceptedWords = Engine::fetchWordsFromFile(wordsFilePath);

    lastFoundWord = std::nullopt;
    lastQueriedSubtring = std::nullopt;
}

std::optional<std::string> GameSession::queryOnSubtring(const std::string& substring)
{
    lastQueriedSubtring = substring;
    ++turnsPlayed;

    lastFoundWord = Engine::playTurn(
        substring,
        acceptedWords,
        letterWeights,
        foundWordCallback
    );

    if (!lastFoundWord.has_value())
    {
        --hearts;
        ++lostHearts;

        if (hearts == 0)
            gameOverCallback();

        return std::nullopt;
    }

    if (allLettersAreUsed())
    {
        resetLetterWeights();
        ++heartRefills;
        hearts = std::min(hearts + 1, maxHearts);
    }

    return lastFoundWord;
}

bool GameSession::allLettersAreUsed()
{
    return std::all_of(
        letterWeights.begin(), letterWeights.end(),
        [](int i) { return i == 0; }
    );
}

void GameSession::resetLetterWeights()
{
    this->letterWeights = Engine::generateDefaultWeights();
}

double GameSession::averageTurnsPerRefill()
{
    return (double)turnsPlayed / (double)heartRefills;
}